tests/returntypes/test_parameter_shift_shot_vec_new.py

```diff
@@ -0,0 +1,2425 @@
+            assert len(j) == len(tape1.trainable_params)
+            assert np.allclose(j[0], 0, atol=shot_vec_tol)
+            assert np.allclose(j[1], exp, atol=shot_vec_tol)
+
+    def test_grad_recipe_parameter_dependent(self, monkeypatch):
+        """Test that an operation with a gradient recipe that depends on
+        its instantiated parameter values works correctly within the parameter
+        shift rule. Also tests that `grad_recipe` supersedes `parameter_frequencies`.
+        """
+
+        class RX(qml.RX):
+            @property
+            def grad_recipe(self):
+                # The gradient is given by [f(2x) - f(0)] / (2 sin(x)), by subsituting
+                # shift = x into the two term parameter-shift rule.
+                x = self.data[0]
+                c = 0.5 / np.sin(x)
+                return ([[c, 0.0, 2 * x], [-c, 0.0, 0.0]],)
+
+        x = np.array(0.654, requires_grad=True)
+        shot_vec = many_shots_shot_vector
+        dev = qml.device("default.qubit", wires=2, shots=shot_vec)
+
+        with qml.tape.QuantumTape() as tape:
+            RX(x, wires=0)
+            qml.expval(qml.PauliZ(0))
+
+        tapes, fn = qml.gradients.param_shift(tape, shots=shot_vec)
+
+        assert len(tapes) == 2
+        assert [t.batch_size for t in tapes] == [None, None]
+        assert qml.math.allclose(tapes[0].operations[0].data[0], 0)
+        assert qml.math.allclose(tapes[1].operations[0].data[0], 2 * x)
+
+        grad = fn(dev.batch_execute(tapes))
+        for g in grad:
+            assert np.allclose(grad, -np.sin(x), atol=shot_vec_tol)
+
+    def test_error_no_diff_info(self):
+        """Test that an error is raised if no grad_recipe, no parameter_frequencies
+        and no generator are found."""
+
+        class RX(qml.RX):
+            """This copy of RX overwrites parameter_frequencies to report
+            missing information, disabling its differentiation."""
+
+            @property
+            def parameter_frequencies(self):
+                raise qml.operation.ParameterFrequenciesUndefinedError
+
+        class NewOp(qml.operation.Operation):
+            """This new operation does not overwrite parameter_frequencies
+            but does not have a generator, disabling its differentiation."""
+
+            num_params = 1
+            grad_method = "A"
+            num_wires = 1
+
+        x = np.array(0.654, requires_grad=True)
+        shot_vec = many_shots_shot_vector
+        dev = qml.device("default.qubit", wires=2, shots=shot_vec)
+
+        for op in [RX, NewOp]:
+            with qml.tape.QuantumTape() as tape:
+                op(x, wires=0)
+                qml.expval(qml.PauliZ(0))
+
+            with pytest.raises(
+                qml.operation.OperatorPropertyUndefined, match="does not have a grad_recipe"
+            ):
+                qml.gradients.param_shift(tape, shots=shot_vec)
+
+
+# TODO: add test class for parameter broadcasting
```

Comment:
is this tracked in a story, along with other TODOs in general? flattening this with the existing test file would be great but I def don't want to try to do that in this PR

Reply:
I've added a bullet point for this. :+1:

There's uncertainty here + likely this is a new feature, hence it was skipped and left as a TODO.

Reply:
sounds great to me!